src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_data_containers.h

```diff
@@ -997,34 +1074,35 @@ namespace sirf {
         void set_csm_smoothness(int s){csm_smoothness_ = s;}
 
         void calculate(CoilImagesVector& iv);
-        void calculate(const MRAcquisitionData& acq)
+        void calculate(MRAcquisitionData& acq)
         {
             CoilImagesVector ci;
             ci.calculate(acq);
             calculate(ci);
         }
 
         CFImage get_csm_as_cfimage(size_t const i) const;
+        CFImage get_csm_as_cfimage(const KSpaceSubset::TagType tag, const int offset) const;
+
 
         void get_dim(size_t const num_csm, int* dim) const
         {
             GadgetronImagesVector::get_image_dimensions(num_csm, dim);
 
         }
 
+        void forward(GadgetronImageData& img, GadgetronImageData& combined_img)const;
```

Comment:
add `const` where you can. Needs to have same order as elsewhere (input vs output)

Reply:
I followed the developer's guide where it says if I pass an output via a pointer or reference it shall come first.
I didn't cross-check it against the rest of the code-base, though.

Reply:
ok great. that should mean that `combined_image` should be a `const GadgetronImageData&`. Maybe you could (in the .h file) name the first arg of `forward` as `output`, just to make it obvious...

Reply:
As @evgueni-ovtchinnikov predicted the problem is that `operator () (x,y,z,channel)` of the `ISMRMRD::Image<T>` only returns `T&` and has no const-overload return to `const T&` . They only provide a method `template <typename T> const T * Image<T>::getDataPtr()` that would allow const access. But to use that to access the data would require to compute a linear index from the arguments `(x,y,z,channel)` called for example `linear_index_x_y_z_c`.

So I see three possibilities:
- I leave it as it is and whenever our image class is involved I can not use  `const`. (Potentially for this PR this would be sufficient.)
- I write the code s.t. when I call the `Image<T>::operator ()` function I compute a linear index and  wherever I use the `operator()` replace that with `data = *(Image::getDataPtr() + linear_index_x_y_z_c)`
- I write a function `const SomeType ImageWrap::operator()` that uses the `getDataPtr()` method and I convert the arguments to a linear index to access the data. (annoying since then for every call you need to get the image header first for the image dimensions) 
- and the 4th possiblity is to add that `const` overload to ISMRMRD, but then we would need to build against their latest version...

Reply:
@johannesmayer: third option can compute the data starting pointer and dimensions in the `ImageWrap` constructor

Reply:
@johannesmayer: perhaps a better idea is to inherit from `ISMRMRD::Image` and add `operator()(x,y,z,c) const` to the derived class

Reply:
I agree, however, I would strongly opt for the 99% instead of the 100% solution here and add a "fix" for this non-constness another time. 

Otherwise, I might have to do a PR from master to my branch one day...

Reply:
Of course, let's not introduce a lot of hard work for ourselves, there's enough already!

Let's take option 1 now, but please create an issue for SIRF that we need this, and one for ISMRMRD, or even do a PR there. At some point they'll merge it, and then we can use it.

By the way, there's a 5th option: use `const_cast`, but it's quite ugly, so I wouldn't do that.

(funny thing is that I can't see where we use `ISMRMRD::Image`)

Reply:
We use it only as a `typedef ISMRMRD::Image<complex_float_t> CFImage` ... 
Then when we get the data the `operator ()` is used, e.g. [here](https://github.com/johannesmayer/SIRF/blob/25393fef1c85b38063e9158ad9d57f19fb1aef92/src/xGadgetron/cGadgetron/fourierencoding.cpp#L104)

Reply:
ok. By the way, if it's ISMRMRD derived code, I wonder if those lines actually are just a straightforward copy, in which case they could be replaced by
```C++
std::copy(img.getDataPtr(), img.getDataPtr()+img.getNumberOfDataElements(), ci.getDataPtr());
```
that wouldn't work if `img.operator()` does some reordering of course.

Reply:
I thought of that already, but then it interfered with
```
Loop nx ny nz nc and do:
dst_img(nx, ny, nz, nc) = coilmap(nx, ny, nz, nc) * ((*ptr_src_img)(nx, ny, nz, 0))
```
so I had dropped it...
But actually that loop problem above is just coil-sensitivity specific, so I think I'll be able to make the Fourier transforms `const` by using your std::copy ...

Reply:
great. Actually
```c++
std::copy(img.begin(). img.end(), ci.getDataPtr());
```
should do the trick I guess. Much neater.

The other example can be done as well with iterators.
```c++
auto coilmap_iter = coilmap.begin();
auto src_img_iter = ptr_src_img->begin();
auto dst_img_iter = dstImg.begin(); // or getDataPtr()
while (coilmap_iter != coilmap_end)
{ ....}
```
or indeed using [std::transform](https://www.cplusplus.com/reference/algorithm/transform/) and [std::multiplies](https://en.cppreference.com/w/cpp/utility/functional/multiplies)
```c++
std::transform(coilmap.begin(), coilmap,end(), ptr_src_image.begin(),
                        dst_img.begin(), // destination
                         std::multiplies<complex_float_t>());
```
or possibly with `foreach` since C++11.

Reply:
Alright looks nice! I'll fiddle that in ...

Reply:
still planning to do this? If not, create an issue for later